### ISSUE TITLE
meaning of `*` `/` `%` `op=` on domains and ranges

### ISSUE THREAD

**Issue by vasslitvinov** [chapel-lang/chapel#19264]
This proposal is to define &nbsp; `*` &nbsp; `/` &nbsp; `%` &nbsp; `op=` &nbsp; on domains and ranges as follows:
* no special handling for &nbsp; `*` &nbsp; `/` &nbsp; `%` ; using these operators with one or two domains/ranges will lead to promotion if possible
* `*=` &nbsp; `/=` &nbsp; `%=` &nbsp; with a domain/range lhs are a compilation error

Discussion: the first rule is a natural choice, as there is no clearly-better alternative and no need to create one. For `op=`, while there is a natural semantics for these operations -- perform the binary operation and store the result -- it is currently unimplemented and is unlikely to be desired. Making these a compilation error for now allows us to enable them later if desired.

**Comment by bradcray** [comment]
> no special handling for   *   /   % ; using these operators with one or two domains/ranges will lead to promotion if possible

If nothing is relying on this behavior currently for ranges or domains, I'd also be encouraging of making these into last-resort compiler errors for the time being as well so that if someone wanted to argue that range*range or rectangular domain*rectangular domain should result in a higher-dimensional domain over time, existing code would not break.

**Comment by vasslitvinov** [comment]
@bradcray what is your motivation for having them be "last resort" ? Are you thinking to allow the user to implement them if desired?

**Comment by bradcray** [comment]
That's right.  "last resort" doesn't thwart overloads, so seems like a best practice for any "I am just here to provide an error" overload.

**Comment by vasslitvinov** [comment]
Makes sense.

Are you thinking to allow users to define general overloads like
```chpl
operator *(a: domain, b) {...}
```
or only more specific ones ex.
```chpl
operator *(a: domain, b: integral) {...}
operator *(a: domain, b) where .... {...}
```

**Comment by bradcray** [comment]
I don't know that I have a specific sense of what a user might want to overload... can you clarify why you're asking?

Your question makes me realize, though, that I had only been imagining providing error overloads for range*range and domain*domain cases.  That said, overloading other cases as errors like `operator *(a: domain, b)` and `operator *(a, b:domain)` may also be attractive now that you've suggested it (?).  At least I don't see any downside if we're not relying on them?

**Comment by vasslitvinov** [comment]
> I don't know that I have a specific sense of what a user might want to overload... can you clarify why you're asking?

Thinking about it again and perhaps answering my own question: a last-resort overload, ex. `operator *(a: domain, b)`, will allow promotion of user-declared overloads, such as `operator *(a:domain, b:int)` when invoked on (domain, array of int). Now that I realize that, I support making them last-resort.

> overloading other cases as errors like operator *(a: domain, b) and operator *(a, b:domain) may also be attractive

Indeed. The experiment that I ran and about to report, defines 8 out of 9 argument combinations (domain, range, any)^2 as errors, see https://github.com/chapel-lang/chapel/commit/6c1516c57a.

**Comment by vasslitvinov** [comment]
Thinking further: if we define the binary ops `*` `/` `%` on domains and ranges to produce compiler errors, should we do the same for `<` `<=` `>=` `>` ?

If you (the reader) think "whoa, this is going too far", then maybe we should also leave alone `*` `/` `%`, allowing them to promote or produce generic "unresolved" error messages? I.e., go for the first rule in the OP.

[Personally I am OK either way if we are consistent across all these operators.]

**Comment by vasslitvinov** [comment]
Reporting on my experimentation.

### Experiments for `+` `-` `|` `&` `^` `*` `/` `%` `op=`

I ran a full paratest on each of the following changes:

* make the above operators (excluding op=) a compiler error if they involve at least one domain or range: https://github.com/vasslitvinov/chapel/commit/6c1516c57a
* make `op=` for the above operators a compiler error if LHS is a domain or a range: https://github.com/vasslitvinov/chapel/compare/f642747985...df4649c85f

### Findings for `*` `/` `%`

* Only a couple of uses as promoted operations. Notable code:
```
// io/ferguson/big-precision-t-lossless
var A:[1..10] real;
A = sqrt(1.0 / (1..10));
```

* No uses of `op=` for these three operators.

### Conclusions for `*` `/` `%` `op=`

* The uses of `*` `/` `%` are rare.
  - This supports turning them into compiler errors.
  - I suggest to do the same for `<` `<=` `>=` `>`.
* There are no uses of `*=` `/=` `%=`.
  - We should turn them into compiler errors.

**Comment by vasslitvinov** [comment]
I opened #19332 to discuss `<` `>` `<=` `>=` on domains, ranges.

**Comment by bradcray** [comment]
> Only a couple of uses as promoted operations. Notable code:
> // io/ferguson/big-precision-t-lossless
> var A:[1..10] real;
> A = sqrt(1.0 / (1..10));

Oh, that's actually kindof cute, and now that I see it, I think I write patterns like this from time to time as well.  So that might suggest either leaving things as-is, or only making the range-range, domain-domain cases into last resort errors.

**Comment by vasslitvinov** [comment]
@bradcray how about patterns like that involving multiplication, like `(1..10) * (1..10)` to produce a stream of squares?

**Comment by bradcray** [comment]
Perhaps, though I still worry that `r1 * r2` would be more useful, and maybe more natural, to interpret as a domain.  That said, I suppose it could be written `{r1, r2}` if someone wanted that.  Interpreting `r ** n` as the n-dimensional domain `{r, r, r}` seems attractive from a rank-neutral programming perspective though, so would it be weird to have `r ** 3` (say) != `r * r * r`?

**Comment by bradcray** [comment]
@vasslitvinov : For this issue, is there a route we could take here in which we just "mark these as unstable" (if there's something to mark) or leave things as-is to defer the question?  Basically I'm looking for ways to prioritize work that seems more important, both in terms of your time (prepping and leading a discussion on this) and others (participating in the discussion).

**Comment by vasslitvinov** [comment]
We are staying with the status quo w.r.t. these operators for 2.0. I removed the 2.0 label.
